crates/apollo-compiler/examples/extract_directives_used_by_query.rs

```diff
@@ -0,0 +1,176 @@
+use apollo_compiler::{
+    hir::{Directive, Field},
+    ApolloCompiler, FileId, HirDatabase,
+};
+
+fn get_directives_used_in_query(compiler: &ApolloCompiler, query_id: &FileId) -> Vec<Directive> {
+    // seed the stack with top-level fields
+    let mut stack: Vec<Field> =
+        compiler
+            .db
+            .operations(*query_id)
+            .iter()
+            .fold(vec![], |mut acc, operation_definition| {
+                acc.extend(operation_definition.selection_set().fields().to_vec());
+                acc
+            });
+
+    let mut directives = vec![];
+
+    // depth first search for nested fields with directives
+    while !stack.is_empty() {
+        if let Some(field) = stack.pop() {
+            if let Some(field_definition) = &field.field_definition(&compiler.db) {
+                directives.extend(field_definition.directives().to_vec());
+            }
+            stack.extend(field.selection_set().fields().to_vec());
+        }
+    }
+
+    directives
+}
+
+fn main() {
+    let schema_src = r#"
+          directive @testDirective(testArg: Boolean!) on FIELD_DEFINITION
+          type GrandChildTest {
+            test: bool @testDirective(testArg: true)
+          }
+          type ChildTest {
+            test: GrandChildTest! @testDirective(testArg: true)
+          }
+          type Test {
+            test: ChildTest! @testDirective(testArg: true)
+          }
+          type Query {
+            testOperation: Test! @testDirective(testArg: true)
+          }
+        "#;
+
+    let mut compiler = ApolloCompiler::new();
+    compiler.add_type_system(schema_src, "not-used-here.graphql");
+
+    let query_src = r#"query {
+          testOperation {
+            test {
+              test {
+                test
+              }
+            }
+          }
+        }
+        "#;
+    let query_id = compiler.add_executable(query_src, "not-used-here.graphql");
+
+    let directives = get_directives_used_in_query(&compiler, &query_id);
+
+    assert_eq!(directives.len(), 4);
```

Comment:
There's room for improvement here of course, `directives` has duplicates present, I couldn't find a simple way to get a hash for uniqueness, the directive location bits differ amongst otherwise duplicate directives. Solving this seems a bit out of scope for a simple example though :)